Adds TensileCreateLibrary cli reference docs (#1924)
ROCm/Tensile: docs/src/cli-reference/TensileCreateLibrary.rst

====================
TensileCreateLibrary
====================

Syntax
------

.. code-block::

    TensileCreateLibrary [OPTIONS...] <LOGIC DIRECTORY> <OUTPUT DIRECTORY> <RUNTIME LANGUAGE>

Required Arguments
------------------

When invoking *TensileCreateLibrary*, the following arguments are required.

\<LOGIC DIRECTORY\>
    Absolute path for logic files. These files are generally found in one of two ways: (i) they are
    generated via the `Tensile` program and placed in the build directory under *3_LibraryLogic* (see :ref:`quick-start-example`).
    (ii) they are found within a project that hosts pre-generated logic files, e.g., `rocBLAS <https://github.com/ROCm/rocBLAS/tree/develop/library/src/blas3/Tensile/Logic>`_.
\<OUTPUT DIRECTORY\>
    Absolute or relative path to the output directory where build artifacts are placed.
\<RUNTIME LANGUAGE\>
    One of: {OCL, HIP, HSA}

Options
-------

When invoking *TensileCreateLibrary*, one can provide zero or more options.

\-\-architecture=ARCHITECTURE
    Architectures to generate a library for. When specifying multiple options, use quoted, semicolon delimited 
    architectures, e.g., --architecture='gfx908;gfx1012'.

>     Supported architechures include: all gfx000 gfx803 gfx900 gfx900:xnack- gfx906 gfx906:xnack+ gfx906:xnack- gfx908 gfx908:xnack+

Reviewer: I think this is my typo but still a typo, should be "architectures"
```suggestion
    Supported architectures include: all gfx000 gfx803 gfx900 gfx900:xnack- gfx906 gfx906:xnack+ gfx906:xnack- gfx908 gfx908:xnack+
```

\-\-embed-library=EMBEDLIBRARY
    Embed (new) library files into static variables. Specify the name of the library.
\-\-embed-library-key=EMBEDLIBRARYKEY
    Access key for embedding library files.
\-\-generate-manifest-and-exit
    Similar to dry-run option for *make*, will compute the outputs
    of *TensileCreateLibrary* and write the expected outputs to a 
    manifest file but does not exectue the commands to generate the 
    output.
\-\-generate-sources-and-exit
    Skip building source code object files and assembly code object files.
    Output source files only and exit. 
\-\-ignore-asm-cap-cache
    Ignore asm capability cache and derive the asm capabilities at runtime.    
\-\-jobs=CPUTHREADS, \-j CPUTHREADS
    Number of parallel jobs to launch. If this options is set higher than *nproc* the number of parallel 
    jobs will be equal to the number of cores. If the this option is set below 1 (e.g. 0 or -1), the number
    of parallel jobs will be set to the number of cores, up to a maximum of 64. (default = -1).    
\-\-lazy-library-loading
    Loads Tensile libraries when needed instead of upfront.
\-\-library-format={yaml,msgpack}

>     Select which library format to use (default: msgpack).

Reviewer: For consistency, let's either change this to `(default = msgpack)` or change L66 above to `(default: -1)`, you can choose.

\-\-library-print-debug

>     Solutions will print enqueue info when enqueueing a kernel.

Reviewer: Enqueue info? Is it worth expanding the meaning of this term?

Author: After taking a longer look at this option, I am not sure that it does anything. We should consider removing.

\-\-no-enumerate
    Do not run rocm_agent_enumerator.
\-\-no-library-print-debug
    Solutions will not print enqueue info when enqueueing a kernel (default).
\-\-no-merge-files
    Store every solution and kernel in separate file.
\-\-no-short-file-names
    Disables short files names.
\-\-num-merged-files=NUMMERGEDFILES
    Number of files the kernels should be written into.
\-\-merge-files
    Store all solutions in single file (default).

> \-\-package-library 
>     Enable the packaging of the client library code objects into separate (deprecated).
>     architectures as sub-directories within the Tensile/library directory.

Reviewer: ```suggestion
\-\-package-library 
    Enable the packaging of the client library code objects into separate
    architectures as sub-directories within the Tensile/library directory (deprecated).
```

Author: Thoughts on noting that undocumented options are not considered supported?

Reviewer: I think it's a good idea, we should also then add `deprecated` to the help docs directly.

\-\-short-file-names

>     On windows kernel names can get too long. 

Reviewer: ```suggestion
    On Windows, kernel names can get too long. 
```

    Converts solution and kernel names to serial ids (default).
\-\-separate-architectures

>     Separates TensileLibrary file by architecture to reduce the time to load the library file.
>     The separate-architectures option writes each architecture into a different TensileLibrary_gfxXXX.dat 
>     file.

Reviewer: ```suggestion
    Separates TensileLibrary file by architecture to reduce the time to load the library file.
    This option writes each architecture into a different TensileLibrary_gfxXXX.dat file.
```

\-\-verbose=PRINTLEVEL, \-v PRINTLEVEL
    Set printout verbosity level {0, 1, 2}.
\-\-version=VERSION
    Version string to embed into library file.
\-\-write-master-solution-index
    Output master solution index in csv format including number 
    of kernels per architecture post build in csv format.

Examples
--------

No options
^^^^^^^^^^

The following command will invoke *TensileCreateLibrary*
with no options passing the Logic directory containing 
logic files and creates a directory *tensile-output* 
in the directory where the *TensileCreateLibrary* 
command was invoked. The *tensile-output* directory
will contain the artifacts.

.. code-block::

    TensileCreateLibrary <absolute path to>/Logic tensile-output HIP

Adding TensileCreateLibrary options 
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The following example illustrates how to add options When

Reviewer: ```suggestion
The following example illustrates how to add options when
```

> invoking *TensileCreateLibrary*. In some cases, such as ``--seprate-architectures``

Reviewer: ```suggestion
invoking *TensileCreateLibrary*. In some cases, such as ``--separate-architectures``
```

no arguments are required; whereas, for ``--jobs`` an argument is required.

.. code-block::

    TensileCreateLibrary --separate-architectures --jobs=32 <absolute path to>Logic tensile-output HIP

.. \-\-cmake-cxx-compiler=CMAKECXXCOMPILER

>     This doesn't appear to do much and I would like to consider removing

Reviewer: Should we delete this from the file, mark as deprecated, or keep it and move it back up to the **Options** section?

Author: Yes I would like to remove it.

    Output master solution index in csv format including number 
    of kernels per architecture post build in csv format.

> Arguments

Author: Can we move these above the options? I feel like the required arguments should come first.

---------

When invoking *TensileCreateLibrary*, the following arguments are required.

\<LOGIC DIRECTORY\>
    Absolute path for logic files. These files are generally found in one of two ways: (i) they are
    generated via the `Tensile` program and placed in the build directory under *3_LibraryLogic* (see :ref:`quick-start-example`).
    (ii) they are found within a project that hosts pre-generated logic files, e.g., `rocBLAS <https://github.com/ROCm/rocBLAS/tree/develop/library/src/blas3/Tensile/Logic>`_.
\<OUTPUT DIRECTORY\>
    Absolute or relative path to the output directory where build artifacts are placed.
\<RUNTIME LANGUAGE\>
    One of: {OCL, HIP, HSA}

> 

Author: I don't know if these qualify as options. We should maybe have a separate section for these.
